Fail when `ecspresso run --dry-run` in v2.4.2

## version
v2.4.2

## Detail

Command:

```bash
$ ecspresso run --dry-run=true
2024/11/02 09:20:31 [INFO] ecspresso version: v2.4.2
2024/11/02 09:20:31 xxx/yyy Running task
2024/11/02 09:20:31 xxx/yyy Task definition ARN: family business-stg-server-app will be registered
2024/11/02 09:20:31 [ERROR] FAILED. failed to describe task definition: operation error ECS: DescribeTaskDefinition, https response error StatusCode: 400, RequestID: business-stg-server-app, ClientException: Family contains invalid characters.
```

But `business-stg-server-app` family is registered.  Also, this command is success using v2.4.1.

Besides, `ecspresso run` is success using v2.4.2.

In the difference of https://github.com/kayac/ecspresso/compare/v2.4.1...v2.4.2 , the following has changed.

- fixes of `run.go`
  - https://github.com/kayac/ecspresso/pull/750
- version of `github.com/aws/aws-sdk-go-v2/xxx` 

Maybe these are related.
Could you please confirm?

--- a/export_test.go
+++ b/export_test.go
@@ -48,8 +48,10 @@ func ResetAWSV2ConfigLoadOptionsFunc() {
 	awsv2ConfigLoadOptionsFunc = nil
 }
 
-func (d *App) TaskDefinitionArnForRun(ctx context.Context, opt RunOption) (string, error) {
-	return d.taskDefinitionArnForRun(ctx, opt)
+type TaskDefinitionForRun = taskDefinitionForRun
+
+func (d *App) ResolveTaskDefinitionForRun(ctx context.Context, opt RunOption) (*taskDefinitionForRun, error) {
+	return d.resolveTaskDefinitionForRun(ctx, opt)
 }
 
 func (opt *DiffOption) SetWriter(w io.Writer) {

--- a/run.go
+++ b/run.go
@@ -71,9 +71,6 @@ func (d *App) Run(ctx context.Context, opt RunOption) error {
 		d.Log("Task definition family %s will be registered", *tdForRun.TaskDefinitionInput.Family)
 	} else {
 		d.Log("Task definition ARN: %s", tdForRun.Arn)
-	}
-
-	if tdForRun.TaskDefinitionInput == nil {
 		var err error
 		td, err := d.DescribeTaskDefinition(ctx, tdForRun.Arn)
 		if err != nil {

--- a/run.go
+++ b/run.go
@@ -63,17 +63,25 @@ func (d *App) Run(ctx context.Context, opt RunOption) error {
 	d.Log("[DEBUG] Overrides")
 	d.LogJSON(ov)
 
-	tdArn, err := d.taskDefinitionArnForRun(ctx, opt)
+	tdForRun, err := d.resolveTaskDefinitionForRun(ctx, opt)
 	if err != nil {
 		return err
 	}
-	d.Log("Task definition ARN: %s", tdArn)
+	if tdForRun.Arn == "" && tdForRun.TaskDefinitionInput != nil {
+		d.Log("Task definition family %s will be registered", *tdForRun.TaskDefinitionInput.Family)
+	} else {
+		d.Log("Task definition ARN: %s", tdForRun.Arn)
+	}
 
-	td, err := d.DescribeTaskDefinition(ctx, tdArn)
-	if err != nil {
-		return err
+	if tdForRun.TaskDefinitionInput == nil {
+		var err error
+		td, err := d.DescribeTaskDefinition(ctx, tdForRun.Arn)
+		if err != nil {
+			return err
+		}
+		tdForRun.TaskDefinitionInput = td
 	}
-	watchContainer := containerOf(td, &opt.WatchContainer)
+	watchContainer := containerOf(tdForRun.TaskDefinitionInput, &opt.WatchContainer)
 	if watchContainer == nil {
 		return fmt.Errorf("container %s not found in the task definition", opt.WatchContainer)
 	}
@@ -84,7 +92,7 @@ func (d *App) Run(ctx context.Context, opt RunOption) error {
 		return nil
 	}
 
-	task, err := d.RunTask(ctx, tdArn, &ov, &opt)
+	task, err := d.RunTask(ctx, tdForRun.Arn, &ov, &opt)
 	if err != nil {
 		return err
 	}
@@ -239,42 +247,47 @@ func (d *App) waitTask(ctx context.Context, task *types.Task, untilRunning bool)
 	return nil
 }
 
-func (d *App) taskDefinitionArnForRun(ctx context.Context, opt RunOption) (string, error) {
+type taskDefinitionForRun struct {
+	Arn                 string
+	TaskDefinitionInput *TaskDefinitionInput
+}
+
+func (d *App) resolveTaskDefinitionForRun(ctx context.Context, opt RunOption) (*taskDefinitionForRun, error) {
 	switch {
 	case *opt.Revision > 0:
 		if opt.LatestTaskDefinition {
-			return "", ErrConflictOptions("revision and latest-task-definition are exclusive")
+			return nil, ErrConflictOptions("revision and latest-task-definition are exclusive")
 		}
 		family, _, err := d.resolveTaskdefinition(ctx)
 		if err != nil {
-			return "", err
+			return nil, err
 		}
-		return fmt.Sprintf("%s:%d", family, *opt.Revision), nil
+		return &taskDefinitionForRun{Arn: fmt.Sprintf("%s:%d", family, *opt.Revision)}, nil
 	case opt.LatestTaskDefinition:
 		family, _, err := d.resolveTaskdefinition(ctx)
 		if err != nil {
-			return "", err
+			return nil, err
 		}
 		d.Log("Revision is not specified. Use latest task definition family " + family)
 		latestTdArn, err := d.findLatestTaskDefinitionArn(ctx, family)
 		if err != nil {
-			return "", err
+			return nil, err
 		}
-		return latestTdArn, nil
+		return &taskDefinitionForRun{Arn: latestTdArn}, nil
 	case opt.SkipTaskDefinition:
 		family, rev, err := d.resolveTaskdefinition(ctx)
 		if err != nil {
-			return "", err
+			return nil, err
 		}
 		if rev != "" {
-			return fmt.Sprintf("%s:%s", family, rev), nil
+			return &taskDefinitionForRun{Arn: fmt.Sprintf("%s:%s", family, rev)}, nil
 		}
 		d.Log("Revision is not specified. Use latest task definition family " + family)
 		latestTdArn, err := d.findLatestTaskDefinitionArn(ctx, family)
 		if err != nil {
-			return "", err
+			return nil, err
 		}
-		return latestTdArn, nil
+		return &taskDefinitionForRun{Arn: latestTdArn}, nil
 	default:
 		tdPath := opt.TaskDefinition
 		if tdPath == "" {
@@ -282,20 +295,20 @@ func (d *App) taskDefinitionArnForRun(ctx context.Context, opt RunOption) (strin
 		}
 		in, err := d.LoadTaskDefinition(tdPath)
 		if err != nil {
-			return "", err
+			return nil, err
 		}
 		{
 			b, _ := MarshalJSONForAPI(in)
 			d.Log("[DEBUG] task definition: %s", string(b))
 		}
 		if opt.DryRun {
-			return fmt.Sprintf("family %s will be registered", *in.Family), nil
+			return &taskDefinitionForRun{Arn: "", TaskDefinitionInput: in}, nil
 		}
 		newTd, err := d.RegisterTaskDefinition(ctx, in)
 		if err != nil {
-			return "", err
+			return nil, err
 		}
-		return *newTd.TaskDefinitionArn, nil
+		return &taskDefinitionForRun{Arn: *newTd.TaskDefinitionArn}, nil
 	}
 }
 

--- a/run_test.go
+++ b/run_test.go
@@ -11,7 +11,8 @@ import (
 
 type taskDefinitionArnForRunSuite struct {
 	opts     []string
-	td       string
+	arn      string
+	family   string
 	raiseErr bool
 }
 
@@ -19,68 +20,72 @@ var testTaskDefinitionArnForRunSuite = map[string][]taskDefinitionArnForRunSuite
 	"tests/run-with-sv.yaml": {
 		{
 			opts: []string{"--skip-task-definition"},
-			td:   "katsubushi:39",
+			arn:  "katsubushi:39",
 		},
 		{
 			opts: []string{"--skip-task-definition", "--revision=42"},
-			td:   "katsubushi:42",
+			arn:  "katsubushi:42",
 		},
 		{
 			opts: []string{"--latest-task-definition"},
-			td:   "katsubushi:45",
+			arn:  "katsubushi:45",
 		},
 		{
 			opts: []string{"--latest-task-definition", "--skip-task-definition"},
-			td:   "katsubushi:45",
+			arn:  "katsubushi:45",
 		},
 		{
 			opts:     []string{"--latest-task-definition", "--skip-task-definition", "--revision=41"},
-			td:       "katsubushi:41",
+			arn:      "katsubushi:41",
 			raiseErr: true, // latest-task-definition and revision are exclusive
 		},
 		{
 			opts:     []string{"--latest-task-definition", "--revision=41"},
-			td:       "katsubushi:41",
+			arn:      "katsubushi:41",
 			raiseErr: true, // latest-task-definition and revision are exclusive
 		},
 		{
-			opts: nil,
-			td:   "family katsubushi will be registered",
+			opts:   nil,
+			arn:    "",
+			family: "katsubushi",
 		},
 		{
-			opts: []string{"--task-def=tests/run-test-td.json"},
-			td:   "family test will be registered",
+			opts:   []string{"--task-def=tests/run-test-td.json"},
+			arn:    "",
+			family: "test",
 		},
 	},
 	"tests/run-without-sv.yaml": {
 		{
 			opts: []string{"--skip-task-definition"},
-			td:   "katsubushi:45",
+			arn:  "katsubushi:45",
 		},
 		{
 			opts: []string{"--skip-task-definition", "--revision=42"},
-			td:   "katsubushi:42",
+			arn:  "katsubushi:42",
 		},
 		{
 			opts: []string{"--latest-task-definition"},
-			td:   "katsubushi:45",
+			arn:  "katsubushi:45",
 		},
 		{
 			opts: []string{"--latest-task-definition", "--skip-task-definition"},
-			td:   "katsubushi:45",
+			arn:  "katsubushi:45",
 		},
 		{
 			opts:     []string{"--latest-task-definition", "--revision=42"},
-			td:       "katsubushi:42",
+			arn:      "katsubushi:42",
 			raiseErr: true, // latest-task-definition and revision are exclusive
 		},
 		{
-			opts: nil,
-			td:   "family katsubushi will be registered",
+			opts:   nil,
+			arn:    "",
+			family: "katsubushi",
 		},
 		{
-			opts: []string{"--task-def=tests/run-test-td.json"},
-			td:   "family test will be registered",
+			opts:   []string{"--task-def=tests/run-test-td.json"},
+			arn:    "",
+			family: "test",
 		},
 	},
 }
@@ -110,7 +115,7 @@ func TestTaskDefinitionArnForRun(t *testing.T) {
 				t.Error(err)
 			}
 			opts := *cliopts.Run
-			tdArn, err := app.TaskDefinitionArnForRun(ctx, opts)
+			td, err := app.ResolveTaskDefinitionForRun(ctx, opts)
 			if s.raiseErr {
 				if err == nil {
 					t.Errorf("%s %s expected error, but got nil", config, args)
@@ -121,8 +126,15 @@ func TestTaskDefinitionArnForRun(t *testing.T) {
 				t.Errorf("%s %s unexpected error: %s", config, args, err)
 				continue
 			}
-			if td := ecspresso.ArnToName(tdArn); td != s.td {
-				t.Errorf("%s %s expected %s, got %s", config, args, s.td, td)
+			if td.Arn != "" {
+				if name := ecspresso.ArnToName(td.Arn); name != s.arn {
+					t.Errorf("%s %s expected %s, got %s", config, args, s.arn, name)
+				}
+			} else {
+				family := *td.TaskDefinitionInput.Family
+				if family != s.family {
+					t.Errorf("%s %s expected %s, got %s", config, args, s.family, family)
+				}
 			}
 		}
 	}

